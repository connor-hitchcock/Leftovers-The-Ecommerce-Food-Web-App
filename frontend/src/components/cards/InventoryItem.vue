<template>
  <v-card max-width="1100px">
    <v-container fluid>
      <v-row>
        <!-- Image column -->
        <v-col align="center" justify="center" cols="auto" md="3" sm="12" style="width: 100%;" v-if="product.images.length === 0">
          <v-icon size="250">
            mdi-image
          </v-icon>
        </v-col>
        <v-col cols="auto" md="3" sm="12" v-else>
          <!-- feed the productImages into the carousel child component -->
          <ProductImageCarousel :productImages="product.images" :showControls="false"/>
        </v-col>

        <!-- Info column -->
        <v-col>
          <v-row>
            <v-col class="auto pa-0" md="10" sm="10">
              <v-card-title
                :class="{ 'pt-0': $vuetify.breakpoint.smAndDown }"
                class="pb-0"
              >
                <!-- shows product name -->
                {{ product.name }}
              </v-card-title>
            </v-col>
            <v-col cols="auto" md="2" sm="2">
              <v-card-actions
                :class="{ 'pt-0': $vuetify.breakpoint.smAndDown }"
                class="pb-0"
              />
            </v-col>
          </v-row>
          <v-row>
            <v-col cols="auto" md="9" sm="12">
              <v-row>
                <!-- shows the product code -->
                <v-card-text
                  :class="{ 'pb-0': $vuetify.breakpoint.mdAndUp }"
                  class="product-fields"
                >
                  <strong>Product Code</strong>
                  <br>
                  {{ product.id }}
                </v-card-text>
              </v-row>
              <v-row>
                <!-- if the description length is more than or equal to 50 without slicing any words, the "Read more..." link will
                appear which will lead the user to the FullProductDescription component  -->
                <template v-if="product.description !== undefined">
                  <span v-if="product.description.length >= 50">
                    <v-card-text
                      id="description"
                      class="pb-0 product-fields"
                    >
                      <strong>Description</strong>
                      <br>
                      {{
                        product.description.replace(
                          /^([\s\S]{50}\S*)[\s\S]*/,
                          "$1"
                        )
                      }}...
                      <!-- feed the productDescription into the dialog box child component -->
                      <FullProductDescription
                        :productDescription="product.description"
                      />
                    </v-card-text>
                  </span>
                  <!-- else just show the product description -->
                  <span v-else>
                    <v-card-text class="pb-0 product-fields">
                      <strong>Description</strong>
                      <br>
                      {{ product.description }}
                    </v-card-text>
                  </span>
                </template>
              </v-row>
              <v-row>
                <!-- shows the product manufacturer -->
                <v-card-text
                  :class="{
                    'pb-0': $vuetify.breakpoint.smAndDown,
                  }"
                  class="product-fields"
                >
                  <strong>Manufacturer</strong>
                  <br>
                  {{ product.manufacturer }}
                </v-card-text>
              </v-row>
              <!-- On small screen merge all non-date info into a single column -->
              <template v-if="$vuetify.breakpoint.xsOnly">
                <v-row>
                  <!-- shows the product price -->
                  <v-card-text class="pb-0 product-fields">
                    <strong>Quantity</strong>
                    <br>
                    {{ inventoryItem.quantity }}
                  </v-card-text>
                </v-row>
                <v-row v-if="inventoryItem.pricePerItem !== undefined">
                  <!-- shows the product price -->
                  <v-card-text class="pb-0 product-fields">
                    <strong>Price per Item</strong>
                    <br>
                    {{ currency.symbol }}{{ inventoryItem.pricePerItem }} {{ currency.code }}
                  </v-card-text>
                </v-row>
                <v-row v-if="inventoryItem.totalPrice !== undefined">
                  <!-- shows the product price -->
                  <v-card-text class="pb-0 product-fields">
                    <strong>Total Price</strong>
                    <br>
                    {{ currency.symbol }}{{ inventoryItem.totalPrice }} {{ currency.code }}
                  </v-card-text>
                </v-row>
              </template>
            </v-col>
            <v-col cols="auto" md="3" sm="12" v-if="$vuetify.breakpoint.smAndUp">
              <v-row>
                <!-- shows the product price -->
                <v-card-text class="pb-0 product-fields">
                  <strong>Quantity</strong>
                  <br>
                  {{inventoryItem.remainingQuantity}}/{{ inventoryItem.quantity }}
                </v-card-text>
              </v-row>
              <v-row v-if="inventoryItem.pricePerItem !== undefined">
                <!-- shows the product price -->
                <v-card-text class="pb-0 product-fields">
                  <strong>Price per Item</strong>
                  <br>
                  {{ currency.symbol }}{{ inventoryItem.pricePerItem }} {{ currency.code }}
                </v-card-text>
              </v-row>
              <v-row v-if="inventoryItem.totalPrice !== undefined">
                <!-- shows the product price -->
                <v-card-text class="pb-0 product-fields">
                  <strong>Total Price</strong>
                  <br>
                  {{ currency.symbol }}{{ inventoryItem.totalPrice }} {{ currency.code }}
                </v-card-text>
              </v-row>
            </v-col>
            <!-- Narrow date view -->
            <v-col v-if="$vuetify.breakpoint.xsOnly" cols="auto" sm="12">
              <v-row>
                <v-card-text class="pb-0 product-fields">
                  <strong>Manufactured</strong>
                  <br>
                  <template v-if="inventoryItem.manufactured === undefined">
                    -
                  </template>
                  <template v-else>
                    {{ formatDate(inventoryItem.manufactured) }}
                  </template>
                </v-card-text>
              </v-row>
              <v-row>
                <v-card-text class="pb-0 product-fields">
                  <strong>Created</strong>
                  <br>
                  <template v-if="inventoryItem.created === undefined">
                    -
                  </template>
                  <template v-else>
                    {{ formatDate(inventoryItem.created) }}
                  </template>
                </v-card-text>
              </v-row>
              <v-row>
                <v-card-text class="pb-0 product-fields">
                  <strong>Sell By</strong>
                  <br>
                  <template v-if="inventoryItem.sellBy === undefined">
                    -
                  </template>
                  <template v-else>
                    {{ formatDate(inventoryItem.sellBy) }}
                  </template>
                </v-card-text>
              </v-row>
              <v-row>
                <v-card-text class="pb-0 product-fields">
                  <strong>Best Before</strong>
                  <br>
                  <template v-if="inventoryItem.bestBefore === undefined">
                    -
                  </template>
                  <template v-else>
                    {{ formatDate(inventoryItem.bestBefore) }}
                  </template>
                </v-card-text>
              </v-row>
              <v-row>
                <v-card-text class="pb-0 product-fields">
                  <strong>Expiry</strong>
                  <br>
                  <template v-if="inventoryItem.expires === undefined">
                    -
                  </template>
                  <template v-else>
                    {{ formatDate(inventoryItem.expires) }}
                  </template>
                </v-card-text>
              </v-row>
            </v-col>
          </v-row>
        </v-col>
        <!-- Timeline column -->
        <v-col cols="auto" >
          <v-timeline
            clipped
            class="timeline"
            v-if="$vuetify.breakpoint.smAndUp"
          >
            <v-timeline-item v-if="inventoryItem.manufactured" small color="green" right>
              <template v-slot:opposite>
                <div class="timeline-label">Manufactured</div>
              </template>
              <div class="timeline-content">
                {{ formatDate(inventoryItem.manufactured) }}
              </div>
            </v-timeline-item>
            <v-timeline-item v-if="product.created" small color="grey" right>
              <template v-slot:opposite>
                <div class="timeline-label">Created</div>
              </template>
              <div class="timeline-content">
                {{ formatDate(product.created) }}
              </div>
            </v-timeline-item>
            <v-timeline-item v-if="inventoryItem.sellBy" small color="yellow" right>
              <template v-slot:opposite>
                <div class="timeline-label">Sell By</div>
              </template>
              <div class="timeline-content">
                {{ formatDate(inventoryItem.sellBy) }}
              </div>
            </v-timeline-item>
            <v-timeline-item v-if="inventoryItem.bestBefore" small color="orange" right>
              <template v-slot:opposite>
                <div class="timeline-label">Best Before</div>
              </template>
              <div class="timeline-content">
                {{ formatDate(inventoryItem.bestBefore) }}
              </div>
            </v-timeline-item>
            <v-timeline-item v-if="inventoryItem.expires" small color="red" f right>
              <template v-slot:opposite>
                <div class="timeline-label">Expiry</div>
              </template>
              <div class="timeline-content">
                {{ formatDate(inventoryItem.expires) }}
              </div>
            </v-timeline-item>
          </v-timeline>
        </v-col>
      </v-row>
      <v-row justify="end">
        <v-tooltip top>
          <template #activator="{on: tooltip}">
            <v-btn
              ref="createSaleItemButton"
              rounded
              outlined
              v-on="tooltip"
              color="primary"
              class="sale-item-but"
              @click="viewCreateSaleItem"
            >
              Create Sale Item
            </v-btn>
          </template>
          <span>Create a Sale from this inventory item</span>
        </v-tooltip>
      </v-row>
    </v-container>
  </v-card>
</template>

<script>
//This component requires two other custom components, one to display the product image, one to view more of the product's description
import FullProductDescription from "../utils/FullProductDescription.vue";
import ProductImageCarousel from "../utils/ProductImageCarousel.vue";
import { currencyFromCountry } from "@/api/currency";
import { formatDate } from '@/utils';

export default {
  name: "InventoryItem",
  props: {
    inventoryItem: Object,
    //retrieved from Inventory page
    businessId: Number
  },
  components: {
    FullProductDescription,
    ProductImageCarousel,
  },
  data() {
    return {
      currency: {},
      showImageUploaderForm: false,
      //If readMoreActivated is false, the product description is less than 50 words, so it wont have to use the FullProductDescription
      //component. Else it will use it and the "Read more..." link will also be shown to lead to the FullProductDescription component
      readMoreActivated: false
    };
  },
  async created() {
    // When the catalogue item is created, the currency will be set to the currency of the country the product is being
    // sold in. It will have blank fields if no currency can be found from the country.
    this.currency = await currencyFromCountry(this.product.countryOfSale);
  },
  computed: {
    /**
     * Helper property for getting the product out of the inventoryItem
     */
    product() {
      return this.inventoryItem.product;
    },
  },
  methods: {
    //if the "Read more..." link if clicked, readMoreActivated becomes true and the FullProductDescription dialog box will open
    activateReadMore() {
      this.readMoreActivated = true;
    },

    placeholderIfEmpty(text) {
      if (text === undefined) {
        return '-';
      }
      return text;
    },

    /**
     * Shows the create Sale Item dialog
     */
    viewCreateSaleItem() {
      this.$store.commit('showCreateSaleItem', {businessId: this.businessId, inventoryItem: this.inventoryItem});
    },
    formatDate,
  },
};
</script>

<style>
.product-fields {
  padding-top: 0;
}

.timeline {
  height: 100%;
}

.timeline-content {
  margin-left: -20px;
}

.timeline-label {
  width: 90px;
  font-weight: bold;
}

.sale-item-but {
  margin: 10px;
  font-weight: bold;
}

</style>
