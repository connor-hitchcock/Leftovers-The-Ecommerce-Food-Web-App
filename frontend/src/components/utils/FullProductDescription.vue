<template>
  <!-- use the v-dialog to show a pop up for the whole product description -->
  <v-dialog v-model="dialog">
    <template v-slot:activator="{ on, attrs }">
      <!-- using the "Read more..." link to activate the pop up dialog -->
      <!--  v-bind="attrs" v-on="on" allows the v-dialog to use this link as the activator for the dialog box -->
      <a
        v-bind="attrs"
        v-on="on"
      >
        Read more
      </a>
    </template>
    <v-card>
      <v-card-title>
        <h2>Description</h2>
      </v-card-title>
      <!-- shows the whole product description -->
      <v-card-text>
        {{ productDescription }}
        <br>
        <v-card-actions class="pa-0">
          <!-- if the user presses return, will close the dialog -->
          <a @click="dialog = false">return</a>
        </v-card-actions>
      </v-card-text>
    </v-card>
  </v-dialog>
</template>

<script>
export default {
  name: "FullProductDescription",
  //pass in productDescription from parent compoenent
  props: ["productDescription"],
  data() {
    return {
      // if dialog is false, the popup does not appear.
      dialog: false,
    };
  },
};
</script>
