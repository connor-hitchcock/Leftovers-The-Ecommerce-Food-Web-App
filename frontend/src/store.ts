import { User, Business, getUser, login, InventoryItem } from './api/internal';
import Vuex, { Store, StoreOptions } from 'vuex';
import { COOKIE, deleteCookie, getCookie, isTesting, setCookie } from './utils';

type UserRole = { type: "user" | "business", id: number };
type SaleItemInfo = { businessId: number, inventoryItem: InventoryItem };

export type StoreData = {
  /**
   * Object representing the current logged in user.
   * If null then no user is logged in
   */
  user: User | null,

  /**
   * The current role the user is acting as.
   * If acting as user then the user is shown content related to their personal account.
   * If acting as a business then the user is shown content related to their business.
   */
  activeRole: UserRole | null,

  /**
   * The global error message that is displayed at the top of the screen.
   *
   * This is the most intruding method for displaying errors and should only be used for errors that
   * must interrupt the application flow.
   * Otherwise error messages should be displayed closer to where they are generated from.
   */
  globalError: string | null,

  /**
   * Whether or not the dialog for registering a business is being shown.
   */
  createBusinessDialogShown: boolean,
  /**
   * The current business the create product dialog is being applied to.
   * If undefined then the create product dialog is hidden.
   */
  createProductDialogBusiness: number | undefined,
  /**
   * Whether or not the dialog for registering a business is being shown.
   */
  createInventoryDialog: number | undefined,
  /**
   * Whether or not the dialog for registering a business is being shown.
   */
  createSaleItemDialog: SaleItemInfo | undefined,
  /**
   * Whether or not the dialog for creating a marketplace card is being shown.
   */
  createMarketplaceCardDialog: User | undefined,
};

function createOptions(): StoreOptions<StoreData> {
  return {
    state: {
      user: null,
      activeRole: null,
      globalError: null,
      createBusinessDialogShown: false,
      createProductDialogBusiness: undefined,
      createInventoryDialog: undefined,
      createSaleItemDialog: undefined,
      createMarketplaceCardDialog: undefined,
    },
    mutations: {
      setUser(state, payload: User) {
        state.user = payload;

        // Ensures that when we log in we always have a role.
        state.activeRole = { type: "user", id: payload.id };

        // If the payload contains a user ID, user is now logged in. Set their session cookie.
        if (payload.id) {
          deleteCookie(COOKIE.USER.toUpperCase());
          deleteCookie(COOKIE.USER.toLowerCase());
          setCookie(COOKIE.USER, payload.id);
        }
      },
      /**
       * Displays an error message at the top of the screen.
       *
       * This is the least perferred method for displaying errors.
       * Since error messages should be displayed closer to where they are generated from.
       *
       * @param state Current state
       * @param error Error message to display
       */
      setError(state, error: string) {
        state.globalError = error;
      },
      /**
       * Dismisses the current error message that is displayed.
       * This function is only expected to be called from the global error message component.
       *
       * @param state Current state
       */
      clearError(state) {
        state.globalError = null;
      },

      logoutUser(state) {
        state.user = null;
        deleteCookie(COOKIE.USER);
      },

      /**
       * Creates a modal create business dialog
       *
       * @param state Current store state
       */
      showCreateBusiness(state) {
        state.createBusinessDialogShown = true;
      },

      /**
       * Hides the create business dialog
       *
       * @param state Current store state
       */
      hideCreateBusiness(state) {
        state.createBusinessDialogShown = false;
      },

      /**
       * Creates a modal create product dialog for adding a product to the provided business
       *
       * @param state Current store state
       * @param businessId Business to create the product for
       */
      showCreateProduct(state, businessId: number) {
        state.createProductDialogBusiness = businessId;
      },

      /**
       * Hides the create product dialog
       *
       * @param state Current store state
       */
      hideCreateProduct(state) {
        state.createProductDialogBusiness = undefined;
      },
      /**
       * Creates a modal create inventory dialog for adding a inventory item to the provided business
       *
       * @param state Current store state
       * @param businessId Business to create the product for
       */
      showCreateInventory(state, businessId: number) {
        state.createInventoryDialog = businessId;
      },
      /**
       * Hides the create inventory dialog
       *
       * @param state Current store state
       */
      hideCreateInventory(state) {
        state.createInventoryDialog = undefined;
      },
      /**
       * Creates a modal create inventory dialog for adding a sale item to the provided business
       *
       * @param state Current store state
       * @param businessId Business to create the sale item for
       */
      showCreateSaleItem(state, saleItemInfo: SaleItemInfo) {
        state.createSaleItemDialog = saleItemInfo;
      },
      /**
       * Hides the create inventory dialog
       *
       * @param state Current store state
       */
      hideCreateSaleItem(state) {
        state.createSaleItemDialog = undefined;
      },

      /**
       * Creates a modal create card dialog for addin a card to the marketplace
       *
       * @param state Current store state
       */
      showCreateMarketplaceCard(state, user : User) {
        state.createMarketplaceCardDialog = user;
      },
      /**
       * Hides the create marketplace card dialog
       *
       * @param state Current store state
       */
      hideCreateMarketplaceCard(state) {
        state.createMarketplaceCardDialog = undefined;
      },

      /**
       * Sets the current user role
       *
       * @param state Current store state
       * @param role Role to act as
       */
      setRole(state, role: UserRole) {
        setCookie('role', JSON.stringify(role));
        state.activeRole = role;
      }
    },
    getters: {
      isLoggedIn(state) {
        return state.user !== null;
      },
      role(state) {
        return state.user?.role;
      },
    },
    actions: {
      /**
       * Attempts to automatically log in the provided user id with the current authentication cookies.
       * Will also set the current role to the previously selected role.
       *
       * @param context The store context
       * @param userId The userId to try to login as
       */
      async autoLogin(context, userId: number) {
        const response = await getUser(userId);
        if (typeof response === 'string') {
          //context.commit('setError', response);
          return;
        }
        context.commit('setUser', response);

        let rawRole = getCookie('role');
        if (rawRole !== null) {
          rawRole = rawRole.split('=')[1];

          const role: UserRole = JSON.parse(rawRole);
          // We should already be logged in at this point, so this should be valid
          const user = context.state.user!;

          if (role.type === 'user') {
            if (user.id === role.id) {
              context.state.activeRole = role;
            } else {
              console.warn('Previous role id does not match current user id');
            }
          } else if (role.type === 'business') {
            let success = false;
            for (const business of (user.businessesAdministered || [])) {
              if (business.id !== role.id) continue;
              context.state.activeRole = role;
              success = true;
              break;
            }
            if (!success) {
              console.warn(`Previous role id does not match a administered business id=${role.id}`);
            }
          } else {
            console.error(`Unknown role type: "${role.type}"`);
          }
        }
      },
      /**
       * Attempts to log in the given user.
       * This will set the cookies and with authenticate future requests.
       *
       * @param context The current context
       * @param Object containing the login credentials
       * @returns Undefined if successful or a string error message
       */
      async login(context, { email, password }) {
        let userId = await login(email, password);
        if (typeof userId === 'string') {
          return userId;
        }
        let user = await getUser(userId);
        if (typeof user === 'string') {
          return user;
        }
        context.commit('setUser', user);

        return undefined;
      }
    }
  };
}


let store: Store<StoreData>;
if (!isTesting()) {
  // If we're in a test enviroment then Vue.use(Vuex) won't have been called yet.
  store = new Vuex.Store(createOptions());
}

/**
 * Resets the global store to the initial state.
 * This function is only to be used in the test enviroment
 */
export function resetStoreForTesting() {
  if (!isTesting()) throw new Error('This function should only be called when testing');
  store = new Vuex.Store(createOptions());
}

/**
 * Gets the global Vuex store.
 *
 * @returns The global Vuex store
 */
export function getStore() {
  return store;
}
