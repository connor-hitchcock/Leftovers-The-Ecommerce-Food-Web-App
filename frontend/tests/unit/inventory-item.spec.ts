
import Vue from 'vue';
import Vuetify from 'vuetify';
import { createLocalVue, mount, Wrapper } from '@vue/test-utils';
import InventoryItem from '@/components/cards/InventoryItem.vue';
// import ProductImageCarousel from "@/components/utils/ProductImageCarousel.vue";
import FullProductDescription from "@/components/utils/FullProductDescription.vue";

import Vuex, { Store } from 'vuex';
import { getStore, resetStoreForTesting, StoreData } from '@/store';

Vue.use(Vuetify);

jest.mock('@/api/currency', () => ({
  currencyFromCountry: jest.fn(() => {
    return {
      code: "test_currency_code",
      symbol: "test_currency_symbol"
    };
  })
}));


describe('InventoryItem.vue', () => {
  let wrapper: Wrapper<any>;
  let vuetify: Vuetify;
  // The global store to be used
  let store: Store<StoreData>;

  /**
   * Set up to test the routing and whether the Product Catalogue item component shows what is required
   */
  beforeEach(async () => {
    const localVue = createLocalVue();
    vuetify = new Vuetify();

    localVue.use(Vuex);
    resetStoreForTesting();
    store = getStore();

    const app = document.createElement ("div");
    app.setAttribute ("data-app", "true");
    document.body.append (app);

    wrapper = mount(InventoryItem, {
      localVue,
      vuetify,
      store,
      components: {
        //ProductImageCarousel will not be tested yet because its part of another task.
        // ProductImageCarousel,
        FullProductDescription
      },
      stubs: {
        //stub ProductImageCarousel because its not related to this test for now.
        //remove stub if testing InventoryItem as a whole
        ProductImageCarousel: true
      },
      mocks: {
        $router: {
          go: () => undefined,
        }
      },
      propsData: {
        inventoryItem: {
          product: {
            name: "test_product_name",
            description: "test_product_description",
            created: "2021-05-12",
            manufacturer: "test_product_manufacturer",
            recommendedRetailPrice: 100,
            id: "test_product_code",
            images: [],
            countryOfSale: "test_country",
          },
          quantity: 4,
          pricePerItem: 7,
          totalPrice: 99,
          manufactured: "2021-05-11",
          sellBy: "2021-05-13",
          bestBefore: "2021-05-14",
          expires: "2021-05-15"
        },
        businessId: 77,
      }
      //Sets up each test case with some values to ensure the Product Catalogue item component works as intended
    });
    await wrapper.setData({
      currency: {
        code: "test_currency_code",
        symbol: "test_currency_symbol"
      },
    });
  });

  it("Must contain the product name", () => {
    expect(wrapper.text()).toContain('test_product_name');
  });

  it("Must contain the product description", () => {
    expect(wrapper.text()).toContain('test_product_description');
  });

  it("Must open dialog box with full product description upon clicking 'Read more...'", async () => {
    await wrapper.setProps({
      inventoryItem: {
        product: {
          name: "test_product_name",
          description: "Some super long description Some super long description Some super long description Some super long description",
          created: "2021-05-12",
          manufacturer: "test_product_manufacturer",
          recommendedRetailPrice: 100,
          id: "test_product_code",
          images: [],
          countryOfSale: "test_country",
        }
      }
    });
    //the description will cut off at the 50th character
    expect(wrapper.text()).toContain(wrapper.vm.product.description.slice(0,50));
    //Full description should not exist
    expect(wrapper.text()).not.toContain(wrapper.vm.product.description);
    Vue.nextTick(() => {
      let productDescriptionComponent = wrapper.findComponent(FullProductDescription);
      //if the component found is not null, means the component exists to be able to read the full description
      expect(productDescriptionComponent).not.toBeNull();
      //value of dialog should be false initially
      expect(productDescriptionComponent.vm.$data.dialog).toBeFalsy();
      //at index 0, the link is the "Read more..." link
      productDescriptionComponent.findAll('a').at(0).trigger("click");
      //wait to let the dialog box load
      Vue.nextTick(() => {
        //now the dialog should be true
        expect(productDescriptionComponent.vm.$data.dialog).toBeTruthy();
        expect(productDescriptionComponent.text()).toContain(wrapper.vm.product.description);
        //at index 1, the link is the "return" link
        productDescriptionComponent.findAll('a').at(1).trigger("click");
        expect(productDescriptionComponent.vm.$data.dialog).toBeFalsy();
      });
    });
  });

  it("Must contain the product manufacturer", () => {
    expect(wrapper.text()).toContain('test_product_manufacturer');
  });

  it("Must contain the product code", () => {
    expect(wrapper.text()).toContain("test_product_code");
  });

  it('Must contain the quantity', () => {
    expect(wrapper.text()).toContain("4");
  });

  it('Must contain the price per item correctly formatted', () => {
    expect(wrapper.text()).toContain("test_currency_symbol7 test_currency_code");
  });

  it('Must contain the total price correctly formatted', () => {
    expect(wrapper.text()).toContain("test_currency_symbol99 test_currency_code");
  });

  it('Must contain manufactured date', () => {
    expect(wrapper.text()).toContain("11 May 2021");
  });

  it('Must contain created date', () => {
    expect(wrapper.text()).toContain("12 May 2021");
  });

  it('Must contain sell by date', () => {
    expect(wrapper.text()).toContain("13 May 2021");
  });

  it('Must contain best before date', () => {
    expect(wrapper.text()).toContain("14 May 2021");
  });

  it('Must contain expiry date', () => {
    expect(wrapper.text()).toContain("15 May 2021");
  });

  it('Must match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Must have a button to create a sale', () => {
    const button = wrapper.findComponent({ref:'createSaleItemButton'});
    expect(button.exists()).toBeTruthy();
  });

  it('Clicking the create sale button opens a dialog to create a sale', async () => {
    const button = wrapper.findComponent({ref:'createSaleItemButton'});
    expect(store.state.createSaleItemDialog).toBeUndefined();
    button.trigger('click');
    await Vue.nextTick();
    expect(store.state.createSaleItemDialog).not.toBeUndefined();
  });
});
