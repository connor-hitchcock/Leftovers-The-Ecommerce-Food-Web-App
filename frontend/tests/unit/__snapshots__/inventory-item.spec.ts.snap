// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InventoryItem.vue Must match snapshot 1`] = `
<div class="v-card v-sheet theme--light" style="max-width: 1100px;">
  <div class="container container--fluid">
    <div class="row">
      <div align="center" justify="center" class="col-sm-12 col-md-3 col-auto" style="width: 100%;"><i aria-hidden="true" class="v-icon notranslate mdi mdi-image theme--light" style="font-size: 250px;"></i></div>
      <div class="col">
        <div class="row">
          <div class="auto pa-0 col-sm-10 col-md-10 col">
            <div class="v-card__title pb-0">
              test_product_name
            </div>
          </div>
          <div class="col-sm-2 col-md-2 col-auto">
            <div class="v-card__actions pb-0"></div>
          </div>
        </div>
        <div class="row">
          <div class="col-sm-12 col-md-9 col-auto">
            <div class="row">
              <div class="v-card__text product-fields pb-0"><strong>Product Code</strong> <br>
                test_product_code
              </div>
            </div>
            <div class="row"><span><div class="v-card__text pb-0 product-fields"><strong>Description</strong> <br>
                    test_product_description
                  </div></span></div>
            <div class="row">
              <div class="v-card__text product-fields"><strong>Manufacturer</strong> <br>
                test_product_manufacturer
              </div>
            </div>
            <!---->
          </div>
          <div class="col-sm-12 col-md-3 col-auto">
            <div class="row">
              <div class="v-card__text pb-0 product-fields"><strong>Quantity</strong> <br>
                /4
              </div>
            </div>
            <div class="row">
              <div class="v-card__text pb-0 product-fields"><strong>Price per Item</strong> <br>
                test_currency_symbol7 test_currency_code
              </div>
            </div>
            <div class="row">
              <div class="v-card__text pb-0 product-fields"><strong>Total Price</strong> <br>
                test_currency_symbol99 test_currency_code
              </div>
            </div>
          </div>
          <!---->
        </div>
      </div>
      <div class="col col-auto">
        <div class="v-timeline timeline theme--light" clipped="">
          <div class="v-timeline-item v-timeline-item--after theme--light">
            <div class="v-timeline-item__body">
              <div class="timeline-content">
                11 May 2021
              </div>
            </div>
            <div class="v-timeline-item__divider">
              <div class="v-timeline-item__dot v-timeline-item__dot--small">
                <div class="v-timeline-item__inner-dot green"></div>
              </div>
            </div>
            <div class="v-timeline-item__opposite">
              <div class="timeline-label">Manufactured</div>
            </div>
          </div>
          <div class="v-timeline-item v-timeline-item--after theme--light">
            <div class="v-timeline-item__body">
              <div class="timeline-content">
                12 May 2021
              </div>
            </div>
            <div class="v-timeline-item__divider">
              <div class="v-timeline-item__dot v-timeline-item__dot--small">
                <div class="v-timeline-item__inner-dot grey"></div>
              </div>
            </div>
            <div class="v-timeline-item__opposite">
              <div class="timeline-label">Created</div>
            </div>
          </div>
          <div class="v-timeline-item v-timeline-item--after theme--light">
            <div class="v-timeline-item__body">
              <div class="timeline-content">
                13 May 2021
              </div>
            </div>
            <div class="v-timeline-item__divider">
              <div class="v-timeline-item__dot v-timeline-item__dot--small">
                <div class="v-timeline-item__inner-dot yellow"></div>
              </div>
            </div>
            <div class="v-timeline-item__opposite">
              <div class="timeline-label">Sell By</div>
            </div>
          </div>
          <div class="v-timeline-item v-timeline-item--after theme--light">
            <div class="v-timeline-item__body">
              <div class="timeline-content">
                14 May 2021
              </div>
            </div>
            <div class="v-timeline-item__divider">
              <div class="v-timeline-item__dot v-timeline-item__dot--small">
                <div class="v-timeline-item__inner-dot orange"></div>
              </div>
            </div>
            <div class="v-timeline-item__opposite">
              <div class="timeline-label">Best Before</div>
            </div>
          </div>
          <div class="v-timeline-item v-timeline-item--after theme--light" f="">
            <div class="v-timeline-item__body">
              <div class="timeline-content">
                15 May 2021
              </div>
            </div>
            <div class="v-timeline-item__divider">
              <div class="v-timeline-item__dot v-timeline-item__dot--small">
                <div class="v-timeline-item__inner-dot red"></div>
              </div>
            </div>
            <div class="v-timeline-item__opposite">
              <div class="timeline-label">Expiry</div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="row justify-end"><button type="button" class="sale-item-but v-btn v-btn--outlined v-btn--rounded theme--light v-size--default primary--text"><span class="v-btn__content">
            Create Sale Item
          </span></button><span class="v-tooltip v-tooltip--top"><!----></span></div>
  </div>
</div>
`;
